builtin/logical/pki/acme_challenges.go

```diff
@@ -0,0 +1,128 @@
+package pki
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	// Size of the raw and url-encoded base64 without padding thumbprint.
+	KeyAuthorizationThumbprintSizeBytes   = 32
+	KeyAuthorizationThumbprintSizeEncoded = 43
+
+	// Size of the raw and url-encoded base64 without padding token.
+	KeyAuthorizationTokenSizeBytes   = 32
+	KeyAuthorizationTokenSizeEncoded = 43
+)
+
+// We take this and apply it to all challenge methods for consistency.
+// Additionally, as the client controls generation of the thumbprint,
+func ValidateKeyAuthorization(keyAuthz string, token string, thumbprint string) (bool, error) {
+	parts := strings.Split(keyAuthz, ".")
+	if len(parts) != 2 {
+		return false, fmt.Errorf("invalid authorization: got %v parts, expected 2", len(parts))
+	}
+
+	tokenPart := parts[0]
+	thumbprintPart := parts[1]
+
+	if token != tokenPart || thumbprint != thumbprintPart {
+		return false, fmt.Errorf("key authorization was invalid")
+	}
+
+	return true, nil
+}
+
+func ValidateHTTP01Challenge(domain string, token string, thumbprint string) (bool, error) {
+	path := "http://" + domain + "/.well-known/acme-challenge/" + token
+
+	transport := &http.Transport{
+		// Only a single request is sent to this server as we do not do any
+		// batching of validation attempts. There is no need to do an HTTP
+		// KeepAlive as a result.
+		DisableKeepAlives:   true,
+		MaxIdleConns:        1,
+		MaxIdleConnsPerHost: 1,
+		MaxConnsPerHost:     1,
+		IdleConnTimeout:     1 * time.Second,
+
+		// We'd rather timeout and re-attempt validation later than hang
+		// too many validators waiting for slow hosts.
+		DialContext: (&net.Dialer{
+			Timeout:   10 * time.Second,
+			KeepAlive: -1 * time.Second,
```

Comment:
We've disabled keep alive above, what does setting this to be negative do?

Reply:
The first is a HTTP keepalive, this is the [TCP keepalive](https://pkg.go.dev/net#Dialer):

```
        // KeepAlive specifies the interval between keep-alive
	// probes for an active network connection.
	// If zero, keep-alive probes are sent with a default value
	// (currently 15 seconds), if supported by the protocol and operating
	// system. Network protocols or operating systems that do
	// not support keep-alives ignore this field.
	// If negative, keep-alive probes are disabled.
	KeepAlive [time](https://pkg.go.dev/time).[Duration](https://pkg.go.dev/time#Duration)
```